brozzler/browser.py

```diff
@@ -494,19 +498,29 @@ def browse_page(
                             'login navigated away from %s; returning!',
                             page_url)
                         self.navigate_to_page(page_url, timeout=page_timeout)
-                behavior_script = brozzler.behavior_script(
-                        page_url, behavior_parameters,
-                        behaviors_dir=behaviors_dir)
-                self.run_behavior(behavior_script, timeout=behavior_timeout)
+                # If the target page HTTP status is 4xx/5xx, there is no point
+                # in running behaviors, outlink and hashtag extraction as we
+                # didn't get a valid page. Screenshot should run because i
+                # may be useful to have a picture of the error page.
+                run_behaviors = True
+                if self.websock_thread.page_status is None or \
+                        self.websock_thread.page_status >= 400:
+                    run_behaviors = False
+
+                if run_behaviors:
+                    behavior_script = brozzler.behavior_script(
+                            page_url, behavior_parameters,
+                            behaviors_dir=behaviors_dir)
+                    self.run_behavior(behavior_script, timeout=behavior_timeout)
+                final_page_url = self.url()
                 if on_screenshot:
                     self._try_screenshot(on_screenshot, screenshot_full_page)
-                if skip_extract_outlinks:
+                if not run_behaviors or skip_extract_outlinks:
                     outlinks = []
                 else:
                     outlinks = self.extract_outlinks()
-                if not skip_visit_hashtags:
-                    self.visit_hashtags(self.url(), hashtags, outlinks)
-                final_page_url = self.url()
```

Comment:
do we want to omit line 509?

Reply:
I moved line 509 to 515 in the new code.

The reason is that in the previous code, we use `self.url()` 2 times a) `self.visit_hashtags(self.url(),...` and b) `final_page_url = self.url()`.
I wanted to use it only once because its slow (it contacts the browser to get the current URL via websocket).
The final page URL doesn't change when you are at that point of the method, so it seems redundant.

So, I execute `final_page_url = self.url()` on line 515 and then I use `final_page_url` instead of `self.url()`.